docs/posts/duckdb-for-rag/index.qmd

```diff
@@ -0,0 +1,543 @@
+```
+
+```{python}
+penguins.sql(sql).limit(3)
+```
+
+## Modified example
+
+```{python}
+sql = text_to_sql(
+    english,
+    table_name=penguins.get_name(),
+    table_schema=penguins.schema(),
+    table_preview=str(penguins),
+)
+ibis.to_sql(penguins.sql(sql))
+```
+
+```{python}
+penguins.sql(sql).limit(3)
+```
+
+:::
+
+Despite high similarity of the English queries in the embedding space, the
+intended results and SQL are very different (descending vs ascending order). If
+you were to retrieve and augment your input to the text-to-SQL function with the
+most similar SQL from past samples, you might confused the model!
+
+Demonstrating this further, let's modify the English query to be equivalent but
+use very different words:
+
+```{python}
+english = """of the birds retrieve the initial ten opposite-of-columns,
+sorted from biggest to smallest by weight"""
+
+s = (
+    t.mutate(similarity=list_cosine_similarity(t.english_embedding, embed(english)))
+    .relocate("similarity")
+    .order_by(ibis._["similarity"].desc())
+    .cache()
+)
+s
+```
+
+In this case, despite resulting in what would be the same generated SQL, the
+cosine similarity of the vectors in their embedding space are much further
+apart.
+
+We can verify that the SQL would be effectively equivalent:
+
+::: {.panel-tabset}
+
+## Initial example
+
+```{python}
+sql = text_to_sql(
+    example_1.english,
+    table_name=penguins.get_name,
+    table_schema=penguins.schema,
+    table_preview=str(penguins),
+)
+ibis.to_sql(penguins.sql(sql))
+```
+
+```{python}
+penguins.sql(sql).limit(3)
+```
+
+## Modified example
+
+```{python}
+sql = text_to_sql(
+    english,
+    table_name=penguins.get_name,
+    table_schema=penguins.schema,
+    table_preview=str(penguins),
+)
+ibis.to_sql(penguins.sql(sql))
+```
+
+```{python}
+penguins.sql(sql).limit(3)
+```
+
+:::
+
+Despite low similarity, these English queries are effectively the same. Using
+our embedding similarity RAG, we might miss this example as it could be under
+the similarity threshold we set. Yet, a LLM can tell that these are semantically
+equivalent, as shown by the SQL it generates for each.
+
+## Discussion
+
+I am by no means an expert on any of this. I studied neural networks to an
+extent in college and I've kept up with all the hype around LLMs, but I'm not a
+researcher nor a practitioner building products.
+
+That said, I have noticed a pattern of tech-debt-inducing code for new machine
+learning technologies when implemented in practice. I hope this blog has
+provoked some thoughts around:
+
+- using bespoke databases when you may be using one that works just fine for the
+ job to be done
+- using embeddings and similarity search for semantic ranking of text
+- alternative approaches for getting the right context into your language models
+
+Also, keep in mind that we're making external API calls to OpenAI for the
+embeddings. This may be costly (though embeddings calls are relatively cheap)
```

Comment:
Might be worth providing an estimate on how much it cost you to run this, so people have a reference.